File: skills/tabletop-dice-rolling/scripts/secure-dice.js
(function installHBTabletopDice(root) {
  'use strict';

  const MODE_ALIASES = Object.freeze({
    cumulative: 'cumulative',
    individual: 'individual total',
    'individual total': 'individual total',
    each: 'individual total',
    'per die': 'individual total'
  });

  const ROLL_RE = /^\s*(?:(?:\/)?roll\s+|r\s+|dice\s+)?(\d*)\s*[dD]\s*(\d+)\s*([+-]\s*\d+)?\s*(cumulative|individual(?:\s+total)?|each|per\s+die)?\s*$/i;

  function requireCrypto() {
    const cryptoObject = root && root.crypto;
    if (!cryptoObject || typeof cryptoObject.getRandomValues !== 'function') {
      throw new Error('A cryptographic RNG is required; weak fallback RNGs are forbidden.');
    }
    return cryptoObject;
  }

  function uniformInt(maxExclusive) {
    if (!Number.isSafeInteger(maxExclusive) || maxExclusive < 1 || maxExclusive > 0x100000000) {
      throw new Error('Invalid random range.');
    }
    const cryptoObject = requireCrypto();
    const range = 0x100000000;
    const limit = Math.floor(range / maxExclusive) * maxExclusive;
    const buf = new Uint32Array(1);
    let value;
    do {
      cryptoObject.getRandomValues(buf);
      value = buf[0];
    } while (value >= limit);
    return value % maxExclusive;
  }

  function parseCommand(command, options = {}) {
    const match = ROLL_RE.exec(String(command || ''));
    if (!match) throw new Error('Unsupported dice command or notation.');
    const count = Number(match[1] || 1);
    const sides = Number(match[2]);
    const modifier = Number(String(match[3] || '0').replace(/\s+/g, ''));
    if (!Number.isInteger(count) || count < 1 || count > 10) throw new Error('Dice count must be between 1 and 10.');
    if (!Number.isInteger(sides) || sides < 2 || sides > 100) throw new Error('Die sides must be between 2 and 100.');
    const rawMode = String(match[4] || options.defaultMode || 'cumulative').toLowerCase().replace(/\s+/g, ' ').trim();
    const mode = MODE_ALIASES[rawMode];
    if (!mode) throw new Error('Mode must be cumulative or individual total.');
    const notation = `${count}d${sides}${modifier ? (modifier > 0 ? `+${modifier}` : `${modifier}`) : ''}`;
    return Object.freeze({ originalCommand: String(command), notation, count, sides, modifier, mode });
  }

  function rollNotation(notation, options = {}) {
    return rollCommand(String(notation), options);
  }

  function rollCommand(command, options = {}) {
    const parsed = parseCommand(command, options);
    const rolls = Object.freeze(Array.from({ length: parsed.count }, () => uniformInt(parsed.sides) + 1));
    const subtotal = rolls.reduce((sum, value) => sum + value, 0);
    const total = subtotal + parsed.modifier;
    const presentation = parsed.mode === 'cumulative'
      ? Object.freeze({ mode: 'cumulative', total })
      : Object.freeze({ mode: 'individual total', results: rolls, ...(parsed.modifier ? { modifier: parsed.modifier } : {}) });
    const spoken = parsed.mode === 'cumulative'
      ? `${parsed.notation} cumulative: ${total}`
      : `${parsed.notation} individual total: ${rolls.join(', ')}${parsed.modifier ? `; aggregate modifier ${parsed.modifier > 0 ? '+' : ''}${parsed.modifier} retained separately` : ''}`;
    return Object.freeze({
      ...parsed,
      rolls,
      subtotal,
      total,
      presentation,
      spoken,
      entropySource: 'web-crypto-getRandomValues-rejection-sampling',
      physicalTrngAttested: false,
      timestamp: new Date().toISOString()
    });
  }

  root.HBTabletopDice = Object.freeze({ parseCommand, rollCommand, rollNotation });
})(typeof globalThis !== 'undefined' ? globalThis : this);

File: skills/tabletop-dice-rolling/scripts/secure_dice.py
#!/usr/bin/env python3
"""Secure tabletop dice roller for HB-TTRPG-tools Agent Skills.

Uses Python's secrets module. Supports 1-10 dice, d2-d100, optional integer
modifier, and presentation modes:
- cumulative: present one summed total
- individual total: present each die in roll order

Cryptographic host randomness is not a physical TRNG claim.
"""
from __future__ import annotations

import argparse
import json
import re
import secrets
from datetime import datetime, timezone

ROLL_RE = re.compile(
    r"^\s*(?:(?:/)?roll\s+|r\s+|dice\s+)?"
    r"(?P<count>\d*)\s*[dD]\s*(?P<sides>\d+)\s*"
    r"(?P<modifier>[+-]\s*\d+)?\s*"
    r"(?P<mode>cumulative|individual(?:\s+total)?|each|per\s+die)?\s*$",
    re.IGNORECASE,
)

MODE_ALIASES = {
    "cumulative": "cumulative",
    "individual": "individual total",
    "individual total": "individual total",
    "each": "individual total",
    "per die": "individual total",
}


def parse_command(command: str, default_mode: str = "cumulative") -> dict:
    match = ROLL_RE.match(command or "")
    if not match:
        raise ValueError("Unsupported dice command or notation.")
    count = int(match.group("count") or 1)
    sides = int(match.group("sides"))
    modifier = int((match.group("modifier") or "0").replace(" ", ""))
    if not 1 <= count <= 10:
        raise ValueError("Dice count must be between 1 and 10.")
    if not 2 <= sides <= 100:
        raise ValueError("Die sides must be between 2 and 100.")
    raw_mode = (match.group("mode") or default_mode).strip().lower()
    mode = MODE_ALIASES.get(raw_mode)
    if mode is None:
        raise ValueError("Mode must be cumulative or individual total.")
    notation = f"{count}d{sides}" + (f"{modifier:+d}" if modifier else "")
    return {
        "original_command": command,
        "notation": notation,
        "count": count,
        "sides": sides,
        "modifier": modifier,
        "mode": mode,
    }


def roll_command(command: str, default_mode: str = "cumulative") -> dict:
    parsed = parse_command(command, default_mode=default_mode)
    rolls = [secrets.randbelow(parsed["sides"]) + 1 for _ in range(parsed["count"])]
    subtotal = sum(rolls)
    total = subtotal + parsed["modifier"]
    if parsed["mode"] == "cumulative":
        presentation = {"mode": "cumulative", "total": total}
        spoken = f"{parsed['notation']} cumulative: {total}"
    else:
        presentation = {"mode": "individual total", "results": rolls}
        if parsed["modifier"]:
            presentation["modifier"] = parsed["modifier"]
        spoken = f"{parsed['notation']} individual total: " + ", ".join(map(str, rolls))
        if parsed["modifier"]:
            spoken += f"; aggregate modifier {parsed['modifier']:+d} retained separately"
    return {
        **parsed,
        "rolls": rolls,
        "subtotal": subtotal,
        "total": total,
        "presentation": presentation,
        "spoken": spoken,
        "entropy_source": "python-secrets-os-cryptographic-rng",
        "physical_trng_attested": False,
        "timestamp": datetime.now(timezone.utc).isoformat(),
    }


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("command", nargs="+", help='Dice command, e.g. "roll 6d10 individual total"')
    parser.add_argument("--default-mode", choices=["cumulative", "individual total"], default="cumulative")
    args = parser.parse_args()
    print(json.dumps(roll_command(" ".join(args.command), args.default_mode), indent=2))


if __name__ == "__main__":
    main()
